### ISSUE TITLE
Convert Blob::Ptr output to a  2D matrix

### ISSUE THREAD

**Issue by baheytharwat** [openvinotoolkit/openvino#7048]
Hi all, I am new to openvino toolkit, I am trying to do model inference using it. I succeeded to convert an onnx model to the IR. 
and I am trying to run this converted model in C++. 
I expect to get an output in the shape of (1, 5000, 3), and I need to have the output in a cv::mat object or a standard 2d matrix, but I am not sure how to access the output from the `Blob::Ptr` type. I use this snippet of code:
```
  Blob::Ptr output = infer_request.GetBlob(output_name);
        
        MemoryBlob::CPtr moutput = as<MemoryBlob>(output);
        if (!moutput) {
            throw std::logic_error("We expect output to be inherited from MemoryBlob, "
                                   "but by fact we were not able to cast output to MemoryBlob");
        }
        // locked memory holder should be alive all time while access to its buffer
        // happens
        auto moutputHolder = moutput->rmap();
        const float* segmentation = moutputHolder.as<const PrecisionTrait<Precision::FP32>::value_type*>(); 
        
        for(int i=0; i<5000; i++){
        	for (int j=0;j<3;j++){
        	 	std::cout << segmentation[j]<< std::endl;
        	 	}
        	 segmentation++;
        	 std::cout<<"---"<<std::endl;
        	 }

        // Print  results
        std::cout << output->size() <<std::endl;

```
I see the printed values are zeros and ones but I expect the output to be logits. So, am I accessing the data correctly? and If yes, why I see that output? Also, If I increase the i value to a value>5000, It works normally and prints zeros and ones as well. although the output should be 5000x3.

Any suggestions?

**Comment by baheytharwat** [comment]
I followed the sample codes in C++ before, and the code I provide is based on these two. 
I believe the model is supported by OpenVINO as I tested the IR representation using python through this code: 
```
from openvino.inference_engine import IECore, Blob, TensorDesc
import numpy as np

XML_PATH = "model.xml"
BIN_PATH = "model.bin"

ie_core_handler = IECore()
network = ie_core_handler.read_network(model=XML_PATH, weights=BIN_PATH)

executable_network = ie_core_handler.load_network(network, device_name='CPU', num_requests=1)

inference_request = executable_network.requests[0]

random_input_data = np.ones((1, 5000, 4)).astype(np.float32)

tensor_description = TensorDesc(precision="FP32", dims=(1, 5000, 4), layout='CHW')
input_blob = Blob(tensor_description, random_input_data)

input_blobs = inference_request.input_blobs

input_blob_name = next(iter(input_blobs))

inference_request.set_blob(blob_name=input_blob_name, blob=input_blob)

inference_request.infer()

output_blob_name = next(iter(inference_request.output_blobs))

output = inference_request.output_blobs[output_blob_name].buffer

print(output[0])
``
It works and gives results same as the onnx model.and here is the output:
![image](https://user-images.githubusercontent.com/54100761/129344819-1230f354-1e34-4b03-9e02-24c60565179a.png)

 My problem is that I am trying to do the inference in C++ and I can't get work with the output correctly `Blob::Ptr`. The C++ code:
```

        const file_name_t input_model = "model.xml";
        const file_name_t input_image_path = "Random point cloud";
        const std::string device_name = "CPU";
        // -----------------------------------------------------------------------------------------------------

        // --------------------------- Step 1. Initialize inference engine core
        // -------------------------------------
        Core ie;
        // -----------------------------------------------------------------------------------------------------

        // Step 2. Read a model in OpenVINO Intermediate Representation (.xml and
        // .bin files) or ONNX (.onnx file) format
        CNNNetwork network = ie.ReadNetwork(input_model);
        if (network.getOutputsInfo().size() != 1)
            throw std::logic_error("Sample supports topologies with 1 output only");
        if (network.getInputsInfo().size() != 1)
            throw std::logic_error("Sample supports topologies with 1 input only");
        // -----------------------------------------------------------------------------------------------------

        // --------------------------- Step 3. Configure input & output
        // ---------------------------------------------
        // --------------------------- Prepare input blobs
        // -----------------------------------------------------
        InputInfo::Ptr input_info = network.getInputsInfo().begin()->second;
        std::string input_name = network.getInputsInfo().begin()->first;

        /* Mark input as resizable by setting of a resize algorithm.
         * In this case we will be able to set an input blob of any shape to an
         * infer request. Resize and layout conversions are executed automatically
         * during inference */
        
//        input_info->getPreProcess().setResizeAlgorithm(RESIZE_BILINEAR);

        input_info->setLayout(Layout::CHW);
        input_info->setPrecision(Precision::FP32);

        // --------------------------- Prepare output blobs
        // ----------------------------------------------------
       
        if (network.getOutputsInfo().empty()) {
            std::cerr << "Network outputs info is empty" << std::endl;
            return EXIT_FAILURE;
        }
        DataPtr output_info = network.getOutputsInfo().begin()->second;
        std::string output_name = network.getOutputsInfo().begin()->first;

        output_info->setPrecision(Precision::FP32);
        // -----------------------------------------------------------------------------------------------------

        // --------------------------- Step 4. Loading a model to the device
        // ------------------------------------------


        ExecutableNetwork executable_network = ie.LoadNetwork(network, device_name);
        // -----------------------------------------------------------------------------------------------------

        // --------------------------- Step 5. Create an infer request
        // -------------------------------------------------


        InferRequest infer_request = executable_network.CreateInferRequest();
        // -----------------------------------------------------------------------------------------------------

        // --------------------------- Step 6. Prepare input
        // --------------------------------------------------------
        /* Read input image to a blob and set it to an infer request without resize
         * and layout conversions. */
        

	cv::Mat image(1, 5000, 4,  CV_32F);
	double low = -100.0;
	double high = +100.0;
	cv::randu(image, cv::Scalar(low), cv::Scalar(high));
	
    	InferenceEngine::TensorDesc tDesc(InferenceEngine::Precision::FP32, {1, 5000, 4}, InferenceEngine::Layout::CHW);

	Blob::Ptr imgBlob = make_shared_blob<float>(tDesc, (float*)image.data);
    	
        // Blob::Ptr imgBlob = wrapMat2Blob(image);     // just wrap Mat data by Blob::Ptr
                                                     // without allocating of new memory

        infer_request.SetBlob(input_name, imgBlob);  // infer_request accepts input blob of any size
        // -----------------------------------------------------------------------------------------------------

        // --------------------------- Step 7. Do inference
        // --------------------------------------------------------
        /* Running the request synchronously */
         infer_request.Infer();
        // -----------------------------------------------------------------------------------------------------

        // --------------------------- Step 8. Process output
        // ------------------------------------------------------
        Blob::Ptr output = infer_request.GetBlob(output_name);
        
        MemoryBlob::CPtr moutput = as<MemoryBlob>(output);
        if (!moutput) {
            throw std::logic_error("We expect output to be inherited from MemoryBlob, "
                                   "but by fact we were not able to cast output to MemoryBlob");
        }
        // locked memory holder should be alive all time while access to its buffer
        // happens
        auto moutputHolder = moutput->rmap();
        const float* segmentation = moutputHolder.as<const PrecisionTrait<Precision::FP32>::value_type*>(); 
        
        InferenceEngine::SizeVector dims = output->getTensorDesc().getDims();
        std::cout<<dims[2]<<std::endl;
        for(int i=0; i<5000; i++){
        	for (int j=0;j<3;j++){
        		//std::cout.precision(30);
        	 	std::cout << segmentation[j]<< std::endl;

        	 	}
        	 segmentation++;
        	 std::cout<<"---"<<std::endl;
        	 }
        std::cout << output->size() <<std::endl;
```
 a sample from the  C++ output:
![image](https://user-images.githubusercontent.com/54100761/129344660-f2c4ed65-d242-421f-b8ad-b38061fdcaa2.png)

so my question is how to convert the blob ptr output to a 2D matrix (PCL point cloud or CV mat, or vector of vectors) in the C++ code? and why the output is different from C++ to the python one?

**Comment by Iffa-Meah** [comment]
Could you share the model you are using?

**Comment by baheytharwat** [comment]
> Could you share the model you are using?

@Iffa-Meah This link contains the onnx model and the converted IR 
https://drive.google.com/drive/folders/1nkIGsaSakqqoyDlhF8TohbJ8PEVvckXB?usp=sharing

**Comment by baheytharwat** [comment]
@Iffa-Meah  any updates from your side?
Do you have any thoughts about the behavior I mentioned?

**Comment by jgespino** [comment]
Hi @baheytharwat 

I apologize for the delay in our response, are you still running into the same issue? If so, could you help us reproduce by providing all the necessary files and commands you used to compile? Also, please include your python code that produces the expected output.

Regards,
Jesus
